mlir/lib/Dialect/EmitC/IR/EmitC.cpp

```diff
@@ -54,6 +58,40 @@ void mlir::emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
   builder.create<emitc::YieldOp>(loc);
 }
 
+bool mlir::emitc::isSupportedEmitCType(Type type) {
+  if (llvm::isa<emitc::OpaqueType>(type))
+    return true;
+  if (auto ptrType = llvm::dyn_cast<emitc::PointerType>(type))
+    return isSupportedEmitCType(ptrType.getPointee());
+  if (auto arrayType = llvm::dyn_cast<emitc::ArrayType>(type)) {
+    auto elemType = arrayType.getElementType();
+    return !llvm::isa<emitc::ArrayType>(elemType) &&
+           isSupportedEmitCType(elemType);
+  }
+  if (type.isIndex())
+    return true;
+  if (llvm::isa<IntegerType>(type))
+    return isSupportedIntegerType(type);
+  if (llvm::isa<FloatType>(type))
+    return isSupportedFloatType(type);
+  if (auto tensorType = llvm::dyn_cast<TensorType>(type)) {
```

Comment:
Is this generalizable to `ShapedType` by chance? If I understand correctly, it's up to a conversion pass to lower the tensors into C-style arrays; so I could see this check passed e.g. by `VectorType`s as well.

Reply:
This function is used to restrict the types on emitc operations, so it actually currently is valid to use tensor types without any conversion beforehand in EmitC (see [here](https://github.com/Xilinx/llvm-project/blob/d288ea348e2d812c5a99eb0af1bf64169d471b43/mlir/lib/Target/Cpp/TranslateToCpp.cpp#L1572) how they will be emitted). This does not hold for vectors, they are unsupported. I went for restricting EmitC operations to accept what the emitter currently supports to not break any current use cases. The benefit of having the changes I propose in this PR is that that we will notice upon creation of the emitc ops that something is broken, instead of only noticing it when trying to emit the invalid code.